src/main/kotlin/org/jitsi/nlj/util/observable/map/ObservableMap.kt

```diff
@@ -0,0 +1,208 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jitsi.nlj.util.observable.map
+
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.function.BiFunction
+import java.util.function.Function
+
+interface MapEventHandler<T, U> {
+    fun entryAdded(newEntry: Map.Entry<T, U>, currentState: Map<T, U>)
+    fun entryUpdated(updatedEntry: Map.Entry<T, U>, currentState: Map<T, U>)
+    fun entryRemoved(removedEntry: Map.Entry<T, U>, currentState: Map<T, U>)
+}
+
+private class MyEntry<T, U>(
+    override val key: T,
+    override val value: U
+) : Map.Entry<T, U>
+
+/**
+ * A [MutableMap] which will notify observers (installed via [onChange]) of any changes to the map
+ * by invoking the given handler with the current state of the map.  It should be noted that some
+ * operations of this Map may be implemented in less efficient methods in order to make
+ * observability simpler.  For example, [ObservableMap.clear] is implemented by removing each
+ * entry one by one.
+ *
+ * NOTE: It was considered to invoke the handlers outside of the lock, but doing so could
+ * result in duplicate/out-of-order events being fired to the handler, so I kept them inside the lock.
+ */
+class ObservableMap<T, U>(private val data: MutableMap<T, U> = mutableMapOf()) : MutableMap<T, U> by data {
+    private val handlers: MutableList<MapEventHandler<T, U>> = CopyOnWriteArrayList()
+    /**
+     * Lock used to make all modifications to the map and notification of handlers
+     * atomic
+     */
+    private val lock = Any()
+
+    fun onChange(handler: MapEventHandler<T, U>) {
+        handlers.add(handler)
+    }
+
+    private fun notifyAdded(key: T, newValue: U) {
+        val dataCopy = data.toMap()
+        handlers.forEach { it.entryAdded(MyEntry(key, newValue), dataCopy) }
+    }
+
+    private fun notifyUpdated(key: T, newValue: U) {
+        val dataCopy = data.toMap()
+        handlers.forEach { it.entryUpdated(MyEntry(key, newValue), dataCopy) }
+    }
+
+    private fun notifyRemoved(key: T, value: U) {
+        val dataCopy = data.toMap()
+        handlers.forEach { it.entryRemoved(MyEntry(key, value), dataCopy) }
+    }
+
+    override fun put(key: T, value: U): U? = synchronized(lock) {
+        return data.put(key, value).also {
+            if (it == null) {
+                notifyAdded(key, value)
+            } else if (it != value) {
+                notifyUpdated(key, value)
+            }
+        }
+    }
```

Comment:
so, i looked at moving notifying the handlers outside of the lock, but  didn't like it for a couple reasons:
1) it was a lot uglier/less clear. the method would look something like this:
```
override fun put(key: T, value: U): U? {
    val (result, event) = synchronized(lock) {
        val result = data.put(key, value)
        val dataCopy = data.toMap()
        val event = if (result ==  null) {
            { notifyAdded(key, value, dataCopy) }
        } else if (result != value) {
             { notifyUpdated(key, value, dataCopy)  }
        }
    }
    event()
    return result
}
```
2) it can result in out of order events, which would be bad. i guess we could add the event to a queue while inside the lock and then drain it outside?

Reply:
With the current use of `ObservableMap` (where we just update a local reference) I think this is fine. I do worry about causing deadlocks if we start to use this elsewhere, but I guess we could fix it with a queue like you suggest without changing the API, so I think we can leave it like this until it becomes a problem.